node/overseer/overseer-gen/proc-macro/src/impl_builder.rs

```diff
@@ -107,46 +76,235 @@ pub(crate) fn impl_builder(info: &OverseerInfo) -> proc_macro2::TokenStream {
+							FieldInit::Fn(fx) =>
+								FieldInit::Fn(Box::new(move |handle: #handle| {
+								let orig = fx(handle)?;
+								Ok(gen_replacement_fn(orig))
+							})),
+							FieldInit::Value(val) =>
+								FieldInit::Value(gen_replacement_fn(val)),
+						};
+						#builder {
+							#fname: replacement,
+							#(
+								#other_subsystem_name: self. #other_subsystem_name,
+							)*
+							#(
+								#baggage_name: self. #baggage_name,
+							)*
+							spawner: self.spawner,
+						}
+					}
+				}
+			}
+		});
+	// Produce setters for all baggage fields as well
+	let baggage_specific_setters = info
+		.baggage()
+		.iter()
+		.enumerate()
+		.map(|(idx, bag_field)| {
+		// Baggage fields follow subsystems
+		let fname = &bag_field.field_name;
+		let ftype = &bag_field.field_ty;
+		let impl_generics = baggage_placeholder_ty[..idx]
+			.iter()
+			.chain(baggage_placeholder_ty[idx + 1..].iter());
+		let other_baggage_name =
+			baggage_name[..idx].iter().chain(baggage_name[idx + 1..].iter());
+
+		let mut uninit_generics = baggage_placeholder_ty
+			.iter()
+			.map(|gen_ty| {
+				let ret : syn::GenericArgument = parse_quote!(#gen_ty);
+				ret
+			})
+			.collect::<Vec<_>>();
+		uninit_generics[idx] = parse_quote!{ FieldMissing<#ftype> };
+		let mut return_type_generics = uninit_generics.clone();
+		return_type_generics[idx] = parse_quote!{ FieldInit<#ftype> };
+		// Baggage can also be generic, so we need to include that to a signature
+		let additional_baggage_generic = if bag_field.generic {
+			quote! {#ftype,}
+		} else {
+			TokenStream::new()
+		};
+
+		quote! {
+			impl <S, #additional_baggage_generic #( #subsystem_placeholder_ty, )* #( #impl_generics, )* >
+			#builder <S, #( #subsystem_placeholder_ty, )* #( #uninit_generics, )* >
+			#builder_where_clause {
+				/// Specify the baggage in the builder
+				pub fn #fname (self, var: #ftype ) ->
+					#builder <S, #( #subsystem_placeholder_ty, )* #( #return_type_generics, )* >
+				{
+					#builder {
+						#fname: FieldInit::<#ftype>::Value(var),
+						#(
+							#subsystem_name: self. #subsystem_name,
+						)*
+						#(
+							#other_baggage_name: self. #other_baggage_name,
+						)*
+						spawner: self.spawner,
+					}
+				}
+			}
+		}
+	});
 
 	let event = &info.extern_event_ty;
 
 	let mut ts = quote! {
-		impl #generics #overseer_name #generics #where_clause {
+		/// Convenience alias.
+		type SubsystemInitFn<T> = Box<dyn FnOnce(#handle) -> ::std::result::Result<T, #error_ty> >;
+		/// Type for the initialised field of the overseer builder
+		pub enum FieldInit<T> {
+			/// Defer initialization to a point where the `handle` is available.
+			Fn(SubsystemInitFn<T>),
+			/// Directly initialize the subsystem with the given subsystem type `T`.
+			/// Also used for baggage fields
+			Value(T),
+		}
+		/// Type marker for the uninitialised field of the overseer builder
+		/// Phantom data is used for type hinting when creating uninitialized
```

Comment:
I'd recommend to use `cargo-spellcheck`, it suggest you to enclose unknown things in `-ticks which is a nice convenience.

Reply:
Having the `PhantomData<T>` really helps to avoid specifying the generics when instantiating the `FooBuilder` when calling `Foo::builder()`, which is very convenient in our use case, imho the effect of something as subtle as this should be named explicitly in the code.